ql/src/java/org/apache/hadoop/hive/ql/PreparePlanUtils.java

```diff
@@ -0,0 +1,263 @@
+    if (plan.getFetchTask() != null) {
+      operators.addAll(getAllFetchOperators(plan.getFetchTask()));
+    }
+    List<Task<?>> allTasks = plan.getRootTasks();
+    List<TezTask> rootTasks = Utilities.getTezTasks(plan.getRootTasks());
+    for(Task task:allTasks) {
+      List<BaseWork> baseWorks = new ArrayList<>();
+      if (task instanceof ExplainTask) {
+        ExplainTask explainTask = (ExplainTask) task;
+        for (Task explainRootTask : explainTask.getWork().getRootTasks()) {
+          if (explainRootTask instanceof TezTask) {
+            TezTask explainTezTask = (TezTask) explainRootTask;
+            baseWorks.addAll(explainTezTask.getWork().getAllWork());
+          }
+        }
+      } else if (task instanceof TezTask) {
+        baseWorks = ((TezTask) task).getWork().getAllWork();
+      }
+      for (BaseWork baseWork : baseWorks) {
+        operators.addAll(baseWork.getAllOperators());
+      }
+    }
+
+    for (Operator<?> op : operators) {
+      switch(op.getType()) {
+      case FILTER:
+        FilterOperator filterOp = (FilterOperator)op;
+        ExprNodeDesc predicate = filterOp.getConf().getPredicate();
+        filterOp.getConf().setPredicate(
+            replaceDynamicParamsWithConstant(predicate, TypeInfoFactory.booleanTypeInfo, parameterMap));
+        break;
+      case SELECT:
+        if (op instanceof VectorSelectOperator) {
+          VectorSelectOperator selectOperator = (VectorSelectOperator) op;
+          List<ExprNodeDesc> selectExprList = selectOperator.getConf().getColList();
+          if (selectExprList != null) {
+            selectOperator.getConf().setColList(replaceDynamicParamsInExprList(selectExprList, parameterMap));
+          }
+        } else {
+          SelectOperator selectOperator = (SelectOperator)op;
+          List<ExprNodeDesc> selectExprList = selectOperator.getConf().getColList();
+          if (selectExprList != null) {
+            selectOperator.getConf().setColList(replaceDynamicParamsInExprList(selectExprList, parameterMap));
+          }
+        }
+        break;
+      default:
+        //exception
+      }
+    }
+  }
+
+  private static List<ExprNodeDesc> replaceDynamicParamsInExprList(List<ExprNodeDesc> exprList,
+      Map<Integer, String> paramMap) {
+    List<ExprNodeDesc> updatedExprList = new ArrayList<>();
+    for (ExprNodeDesc expr:exprList) {
+      expr = replaceDynamicParamsWithConstant(expr, expr.getTypeInfo(), paramMap);
+      updatedExprList.add(expr);
+    }
+    return updatedExprList;
+  }
+
+  // Note about type inference
+  // Since dynamic parameter lacks type we need to figure out appropriate type to create constant
+  // out of string value. To do this, we choose the type of first child of the parent expression
+  // which isn't dynamic parameter
+  // TODO: cases to consider/cover
+  //  exprs have no children, expres have all dynamic parameters
+  private static ExprNodeDesc replaceDynamicParamsWithConstant(ExprNodeDesc expr,
```

Comment:
This is a difficult problem indeed. I guess this assumption is fine right now but it seems it can cause different behavior compared to not using prepare statement. We should be clear about this and discuss how we will do inference moving forward (we should check the standard to see how they deal with it). In addition, I am wondering whether, for the time being, we should limit dynamic parameters to filter expressions because of this (it seems replacing for project may have more side effects over the plan).